### DISCUSSION TITLE
How can we show a visualization of the status of a reconciliation?

### DISCUSSION THREAD

**Discussion by jpellizzari** [fluxcd/flux2#1431] [Proposals] [answered]
It would be great to show a visualization of the state of a reconciliation part of the [Web UI](https://github.com/fluxcd/webui) project. I wanted to start a discussion on what work would need to be done to be able to surface the data to render such a view, either as part of the Web UI, or possibly in the CLI as well.

The basic user flow would be as follows:

1. User clicks the "Sync" button or does a `flux reconcile` command
2. The progress/state of the different controllers and/or kubernetes objects is visible in a central user-interface (browser, CLI, or otherwise)
3. A final completion or error state can be discerned

I am imagining a directed-graph style view to show all of the affected objects, but a linear or tabular view might work as well.

**Comment by squaremo** [comment]
Some notes that may be obvious or not, to get us started ..

 * To show what happened when syncing a particular commit, you have to know what was applied in the commit. Annotations/Labels added by the GOTK controllers might help with that.
 * Once you can do that, you might be able to go back and look at the last sync, or just the current state of things synced. GOTK is oriented around syncs happening automatically -- and, even if someone clicks a "sync" button, that doesn't mean the sync for the latest commit hasn't already been done -- so this is a more fundamental requirement.
 * Generally speaking, an object with a "controller" owner indicates a child relation; e.g., ReplicaSet to Deployment. But given the parent (and the Kubernetes API), to work efficiently you would have to know what kind of child to look for.
 * kstatus should be good enough to give an answer from `{progressing, stalled, ready}`

**Reply by jpellizzari** [reply]
@squaremo 
> To show what happened when syncing a particular commit, you have to know what was applied in the commit. Annotations/Labels added by the GOTK controllers might help with that.

Is there a way to generically scan for these objects without knowing their `Kind`?

**Reply by jpellizzari** [reply]
I was looking at the `RESTMapper` package for this type of functionality ^^. Not sure if I am on the right track.

**Reply by squaremo** [reply]
You can look up all the kinds in a cluster via the discovery API, but it's pretty inefficient. Happily, each Kustomization object keeps a catalogue of the namespaces and kinds of objects it synced, so you can target exact kinds.

**Reply by squaremo** [reply]
.. and you can use `"controller-runtime/pkg/client".Client` by passing `*"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured".Unstructured` values with the appropriate GroupVersionKind set.

**Reply by jpellizzari** [reply]
@squaremo My client is complaining about certain `Kinds` not being present in the `scheme`. Any tips to avoid listing every object in my `scheme`?

**Reply by jpellizzari** [reply]
Fixed my issue for now by doing:
```go
import  autoscalingv2beta2 "k8s.io/api/autoscaling/v2beta2"
//...
_ = autoscalingv2beta2.AddToScheme(scheme)
```
Unsure if I need to add every possible resource type to my scheme, since Flux can reconcile literally everything?

**Reply by jpellizzari** [reply]
Nevermind, I think was using an `unstructured.Unstructured` instead of an `unstructure.UnstructuredList`.

**Comment by jpellizzari** [comment]
Made a boatload of progress thanks to the input from @squaremo.

> Generally speaking, an object with a "controller" owner indicates a child relation; e.g., ReplicaSet to Deployment. But given the parent (and the Kubernetes API), to work efficiently you would have to know what kind of child to look for.

I am going to maintain a lookup of "Known Children" to handle commonly used objects in lieu of a better idea (starting with Deployment -> ReplicaSet -> Pod).

The next hurdle will be figuring out how to query the by the `ownershipReferences` without resorting to getting a list of everything and filtering via a `for` loop. @hiddeco @squaremo @stefanprodan @stealthybox Any insight here would be very appreciated.

**Reply by hiddeco** [reply]
> The next hurdle will be figuring out how to query the by the `ownershipReferences` without resorting to getting a list of everything and filtering via a `for` loop. 

The Kubernetes API does not allow this, but for resources that have pruning enabled (until https://github.com/fluxcd/kustomize-controller/issues/305 is solved), you can query the kinds/namespaces with a label selector.

**Reply by stefanprodan** [reply]
@hiddeco no matter the prune flag the name/namespace labels are there.

**Reply by hiddeco** [reply]
I stand corrected.

**Comment by jpellizzari** [comment, marked as answer]
Made some progress: 
![Screenshot from 2021-05-25 16-17-34](https://user-images.githubusercontent.com/2802257/119580176-c55d1580-bd74-11eb-8393-666f1007da6f.png)

Just needs to a fresh coat of paint (in the form of a fancy directed-graph view). I think the rest of this issue will be discussed via PR, so marking as answered.
